packages/embark/src/lib/modules/ipfs/index.js

```diff
@@ -98,60 +128,74 @@ class IPFS {
   }
 
   addStorageProviderToEmbarkJS() {
-    this.events.request('version:downloadIfNeeded', 'ipfs-api', (err, location) => {
-      if (err) {
-        this.logger.error(__('Error downloading IPFS API'));
-        return this.logger.error(err.message || err);
-      }
-      this.events.request('code-generator:ready', () => {
-        this.events.request('code-generator:symlink:generate', location, 'ipfs-api', (err, _symlinkDest) => {
-          if (err) {
-            this.logger.error(__('Error creating a symlink to IPFS API'));
-            return this.logger.error(err.message || err);
-          }
-
-          this.events.emit('runcode:register', 'IpfsApi', require('ipfs-api'), () => {
-            let code = "";
-            code += "\nconst __embarkIPFS = require('embarkjs-ipfs')";
-            code += "\nEmbarkJS.Storage.registerProvider('ipfs', __embarkIPFS.default || __embarkIPFS);";
-
-            this.embark.addCodeToEmbarkJS(code);
-            this.embark.addConsoleProviderInit("storage", code, (storageConfig) => storageConfig.enabled);
+    if(!this.addedToEmbarkJs) {
+      this.addedToEmbarkJs = true;
+      this.events.request('version:downloadIfNeeded', 'ipfs-api', (err, location) => {
+        if (err) {
+          this.logger.error(__('Error downloading IPFS API'));
+          return this.logger.error(err.message || err);
+        }
+        this.events.request('code-generator:ready', () => {
+          this.events.request('code-generator:symlink:generate', location, 'ipfs-api', (err, _symlinkDest) => {
+            if (err) {
+              this.logger.error(__('Error creating a symlink to IPFS API'));
+              return this.logger.error(err.message || err);
+            }
+
+            this.events.emit('runcode:register', 'IpfsApi', require('ipfs-api'), () => {
+              let code = "";
+              code += "\nconst __embarkIPFS = require('embarkjs-ipfs')";
+              code += "\nEmbarkJS.Storage.registerProvider('ipfs', __embarkIPFS.default || __embarkIPFS);";
+
+              this.embark.addCodeToEmbarkJS(code);
+              this.embark.addConsoleProviderInit("storage", code, (storageConfig) => storageConfig.enabled);
+            });
           });
         });
       });
-    });
+    }
   }
 
   addObjectToConsole() {
-    const {host, port} = this._getNodeUrlConfig();
-    let ipfs = IpfsApi(host, port);
-    this.events.emit("runcode:register", "ipfs", ipfs);
+    if(!this.addedToConsole) {
```

Comment:
Reverse and early return here too please.

Reply:
Updated.